File: packages/sdk/java/src/main/java/com/ihui/ai/sdk/module/FilesApi.java
package com.ihui.ai.sdk.module;

import com.fasterxml.jackson.databind.JsonNode;
import com.ihui.ai.sdk.BaseClient;
import com.ihui.ai.sdk.model.UploadInitRequest;

import java.io.File;
import java.util.Map;

import okhttp3.MultipartBody;
import okhttp3.RequestBody;

/**
 * 文件模块 — 列表 / 上传 / 详情 / 删除 / 内容 / 版本 / 分片上传。
 *
 * <p>端点(9 个):
 * <ul>
 *   <li>GET  /v1/files(文件列表)</li>
 *   <li>POST /v1/files(上传文件,multipart/form-data)</li>
 *   <li>GET  /v1/files/:id(文件详情)</li>
 *   <li>DELETE /v1/files/:id(删除文件)</li>
 *   <li>GET  /v1/files/:id/content(文件内容,二进制流)</li>
 *   <li>GET  /v1/files/:id/versions(文件版本)</li>
 *   <li>POST /v1/files/upload-init(分片上传初始化)</li>
 *   <li>POST /v1/files/upload-chunk(上传分片)</li>
 *   <li>POST /v1/files/complete(完成上传)</li>
 * </ul>
 */
public final class FilesApi {

    private final BaseClient client;

    /**
     * 构造 FilesApi。
     *
     * @param client 底层 BaseClient
     */
    public FilesApi(BaseClient client) {
        this.client = client;
    }

    /**
     * GET /v1/files(文件列表)。
     *
     * @return 文件列表(JsonNode)
     */
    public JsonNode list() {
        return client.request("GET", "/files", null, JsonNode.class);
    }

    /**
     * POST /v1/files(上传文件,multipart/form-data)。
     *
     * @param file 本地文件
     * @return 上传结果(JsonNode,含 fileId / filename / bytes)
     */
    public JsonNode upload(File file) {
        return upload(file, file.getName());
    }

    /**
     * POST /v1/files(上传文件,multipart/form-data,自定义文件名)。
     *
     * @param file     本地文件
     * @param filename 上传文件名
     * @return 上传结果(JsonNode)
     */
    public JsonNode upload(File file, String filename) {
        RequestBody fileBody = RequestBody.create(file,
                okhttp3.MediaType.parse("application/octet-stream"));
        MultipartBody body = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("file", filename, fileBody)
                .build();
        return client.requestMultipart("/files", body, JsonNode.class);
    }

    /**
     * POST /v1/files(上传字节数组,multipart/form-data)。
     *
     * @param data     文件字节
     * @param filename 文件名
     * @param mimeType MIME 类型(如 image/png;可为 null)
     * @return 上传结果(JsonNode)
     */
    public JsonNode upload(byte[] data, String filename, String mimeType) {
        okhttp3.MediaType mt = mimeType != null
                ? okhttp3.MediaType.parse(mimeType)
                : okhttp3.MediaType.get("application/octet-stream");
        RequestBody fileBody = RequestBody.create(data, mt);
        MultipartBody body = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("file", filename, fileBody)
                .build();
        return client.requestMultipart("/files", body, JsonNode.class);
    }

    /**
     * GET /v1/files/:id(文件详情)。
     *
     * @param id 文件 ID
     * @return 文件详情(JsonNode)
     */
    public JsonNode get(String id) {
        return client.request("GET", "/files/" + BaseClient.encode(id), null, JsonNode.class);
    }

    /**
     * DELETE /v1/files/:id(删除文件)。
     *
     * @param id 文件 ID
     */
    public void delete(String id) {
        client.requestRaw("DELETE", "/files/" + BaseClient.encode(id), null);
    }

    /**
     * GET /v1/files/:id/content(文件内容,返回二进制字节)。
     *
     * @param id 文件 ID
     * @return 文件内容字节
     */
    public byte[] getContent(String id) {
        return client.requestBytes("/files/" + BaseClient.encode(id) + "/content");
    }

    /**
     * GET /v1/files/:id/versions(文件版本列表)。
     *
     * @param id 文件 ID
     * @return 版本列表(JsonNode)
     */
    public JsonNode getVersions(String id) {
        return client.request("GET", "/files/" + BaseClient.encode(id) + "/versions",
                null, JsonNode.class);
    }

    /**
     * POST /v1/files/upload-init(分片上传初始化)。
     *
     * @param req 初始化请求
     * @return 初始化结果(JsonNode,含 uploadId / chunkSize)
     */
    public JsonNode uploadInit(UploadInitRequest req) {
        return client.request("POST", "/files/upload-init", req, JsonNode.class);
    }

    /**
     * POST /v1/files/upload-chunk(上传分片,multipart)。
     *
     * @param uploadId  上传 ID(来自 upload-init)
     * @param index     分片索引(0-based)
     * @param chunkData 分片字节
     * @return 上传结果(JsonNode)
     */
    public JsonNode uploadChunk(String uploadId, int index, byte[] chunkData) {
        RequestBody chunkBody = RequestBody.create(chunkData,
                okhttp3.MediaType.get("application/octet-stream"));
        MultipartBody body = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("uploadId", uploadId)
                .addFormDataPart("index", String.valueOf(index))
                .addFormDataPart("chunk", "chunk-" + index, chunkBody)
                .build();
        return client.requestMultipart("/files/upload-chunk", body, JsonNode.class);
    }

    /**
     * POST /v1/files/complete(完成分片上传)。
     *
     * @param req 请求体(uploadId 等)
     * @return 完成结果(JsonNode,含 fileId / status)
     */
    public JsonNode uploadComplete(Map<String, Object> req) {
        return client.request("POST", "/files/complete", req, JsonNode.class);
    }
}

File: packages/sdk/dotnet/Modules/FilesApi.cs
using System.Net.Http.Headers;
using System.Text.Json;

namespace Ihui.AI;

/// <summary>
/// 文件模块 — 列表 / 上传 / 详情 / 删除 / 内容 / 版本 / 分片上传。
/// </summary>
/// <remarks>
/// 端点(9 个):
/// <list type="bullet">
///   <item>GET  /v1/files(文件列表)</item>
///   <item>POST /v1/files(上传文件,multipart/form-data)</item>
///   <item>GET  /v1/files/:id(文件详情)</item>
///   <item>DELETE /v1/files/:id(删除文件)</item>
///   <item>GET  /v1/files/:id/content(文件内容,二进制流)</item>
///   <item>GET  /v1/files/:id/versions(文件版本)</item>
///   <item>POST /v1/files/upload-init(分片上传初始化)</item>
///   <item>POST /v1/files/upload-chunk(上传分片)</item>
///   <item>POST /v1/files/upload-complete(完成上传)</item>
/// </list>
/// </remarks>
public sealed class FilesApi
{
    private readonly BaseClient _client;

    internal FilesApi(BaseClient client) => _client = client;

    /// <summary>GET /v1/files(文件列表)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>文件列表(JsonElement)</returns>
    public Task<JsonElement?> ListAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/files", null, cancellationToken);

    /// <summary>POST /v1/files(上传文件流,multipart/form-data)。</summary>
    /// <param name="stream">文件流</param>
    /// <param name="filename">文件名</param>
    /// <param name="mimeType">MIME 类型(可选,默认 application/octet-stream)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>上传结果(JsonElement,含 fileId / filename / bytes)</returns>
    public Task<JsonElement?> UploadAsync(Stream stream, string filename, string? mimeType = null, CancellationToken cancellationToken = default)
    {
        var content = new MultipartFormDataContent();
        var byteContent = new StreamContent(stream);
        byteContent.Headers.ContentType = new MediaTypeHeaderValue(mimeType ?? "application/octet-stream");
        content.Add(byteContent, "file", filename);
        return _client.RequestMultipartAsync<JsonElement>("/files", content, cancellationToken)!;
    }

    /// <summary>POST /v1/files(上传字节数组,multipart/form-data)。</summary>
    /// <param name="data">文件字节</param>
    /// <param name="filename">文件名</param>
    /// <param name="mimeType">MIME 类型(可选)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>上传结果(JsonElement)</returns>
    public Task<JsonElement?> UploadAsync(byte[] data, string filename, string? mimeType = null, CancellationToken cancellationToken = default)
    {
        var content = new MultipartFormDataContent();
        var byteContent = new ByteArrayContent(data);
        byteContent.Headers.ContentType = new MediaTypeHeaderValue(mimeType ?? "application/octet-stream");
        content.Add(byteContent, "file", filename);
        return _client.RequestMultipartAsync<JsonElement>("/files", content, cancellationToken)!;
    }

    /// <summary>GET /v1/files/:id(文件详情)。</summary>
    /// <param name="id">文件 ID</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>文件详情(JsonElement)</returns>
    public Task<JsonElement?> GetAsync(string id, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/files/" + BaseClient.Encode(id), null, cancellationToken);

    /// <summary>DELETE /v1/files/:id(删除文件)。</summary>
    /// <param name="id">文件 ID</param>
    /// <param name="cancellationToken">取消令牌</param>
    public Task DeleteAsync(string id, CancellationToken cancellationToken = default)
        => _client.RequestRawAsync("DELETE", "/files/" + BaseClient.Encode(id), null, cancellationToken)!;

    /// <summary>GET /v1/files/:id/content(文件内容,返回二进制字节)。</summary>
    /// <param name="id">文件 ID</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>文件内容字节</returns>
    public Task<byte[]> GetContentAsync(string id, CancellationToken cancellationToken = default)
        => _client.RequestBytesAsync("/files/" + BaseClient.Encode(id) + "/content", cancellationToken);

    /// <summary>GET /v1/files/:id/versions(文件版本列表)。</summary>
    /// <param name="id">文件 ID</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>版本列表(JsonElement)</returns>
    public Task<JsonElement?> GetVersionsAsync(string id, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/files/" + BaseClient.Encode(id) + "/versions", null, cancellationToken);

    /// <summary>POST /v1/files/upload-init(分片上传初始化)。</summary>
    /// <param name="req">初始化请求</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>初始化结果(JsonElement,含 uploadId / chunkSize)</returns>
    public Task<JsonElement?> UploadInitAsync(UploadInitRequest req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/files/upload-init", req, cancellationToken);

    /// <summary>POST /v1/files/upload-chunk(上传分片,multipart)。</summary>
    /// <param name="uploadId">上传 ID(来自 upload-init)</param>
    /// <param name="index">分片索引(0-based)</param>
    /// <param name="chunkData">分片字节</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>上传结果(JsonElement)</returns>
    public Task<JsonElement?> UploadChunkAsync(string uploadId, int index, byte[] chunkData, CancellationToken cancellationToken = default)
    {
        var content = new MultipartFormDataContent();
        content.Add(new StringContent(uploadId), "uploadId");
        content.Add(new StringContent(index.ToString()), "index");
        var chunkContent = new ByteArrayContent(chunkData);
        chunkContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
        content.Add(chunkContent, "chunk", "chunk-" + index);
        return _client.RequestMultipartAsync<JsonElement>("/files/upload-chunk", content, cancellationToken)!;
    }

    /// <summary>POST /v1/files/upload-complete(完成分片上传)。</summary>
    /// <param name="req">请求体(uploadId 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>完成结果(JsonElement,含 fileId / status)</returns>
    public Task<JsonElement?> UploadCompleteAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/files/upload-complete", req, cancellationToken);
}
